internal/configs/virtualserver.go

```diff
@@ -81,9 +81,22 @@ func (namer *variableNamer) GetNameForVariableForRulesRouteMainMap(rulesIndex in
 	return fmt.Sprintf("$vs_%s_rules_%d", namer.safeNsName, rulesIndex)
 }
 
+func getUpstream(upstreamName string, upstreams []conf_v1alpha1.Upstream) conf_v1alpha1.Upstream {
+	for _, u := range upstreams {
+		if u.Name == upstreamName {
+			return u
+		}
+	}
+	return conf_v1alpha1.Upstream{}
+}
+
 func generateVirtualServerConfig(virtualServerEx *VirtualServerEx, tlsPemFileName string, baseCfgParams *ConfigParams, isPlus bool) version2.VirtualServerConfig {
 	ssl := generateSSLConfig(virtualServerEx.VirtualServer.Spec.TLS, tlsPemFileName, baseCfgParams)
 
+	// oUpstreams keeps track of conf_v1alpha1.Upstreams as they are created
+	// necessary to know what Location references which Upstream
+	var oUpstreams []conf_v1alpha1.Upstream
```

Comment:
use a map to store upstreams. this way you won't need to have `getUpstream` function.

the key of the map should be the generated upstream name. That name is unique among all upstream. Right now there is a bug in your code: consider the case when you have an upstream with the same name  in VS and VSR.

regarding the variable name, I suggest calling it `crUpstreams`, where cr stands fo custom resources.